core/embed/lib/error_handling.h

```diff
@@ -20,6 +20,167 @@
 #ifndef LIB_ERROR_HANDLING_H
 #define LIB_ERROR_HANDLING_H
 
+// Suppresses the intellisense error in VSCode
+#ifndef __FILE_NAME__
+#define __FILE_NAME__ __FILE__
+#endif
+
+// Status code type
+typedef struct {
+  // Do not access this field directly,
+  // use `ts_ok()` and `ts_error()` macros.
+  uint32_t code;
+} ts_t;
+
+// Build status code from any 16-bit value.
+//
+// Status codes are hardened against fault injections
+// by storing the same value in the upper 16 bits.
+#define TS_BUILD(code) ((ts_t){(code) | ((code) << 16)})
+
+// OK status code (signalling success or no error)
+#define TS_OK TS_BUILD(0)
+
+// This offset ensures at lest 16bit hamming distance
+// between `TS_OK` and other status codes.
+#define TS_ERROR_OFFSET 0xFF00
+
+// Build error status code from any 8-bit value
+//
+// The code should be in range 0 to 255 to ensure
+// 16-bit hamming distance to `TS_OK`
+#define TS_ERROR_BUILD(code) TS_BUILD(TS_ERROR_OFFSET + (code))
+
+// Error status codes
+#define TS_ERROR TS_ERROR_BUILD(0)
+#define TS_ERROR_BUSY TS_ERROR_BUILD(1)
+#define TS_ERROR_TIMEOUT TS_ERROR_BUILD(2)
+#define TS_ERROR_NOTINIT TS_ERROR_BUILD(3)
+#define TS_ERROR_ARG TS_ERROR_BUILD(4)
+#define TS_ERROR_IO TS_ERROR_BUILD(5)
+
+// Check status code consistency and returns its value.
+// If invalid status code is detected, it will call `__fatal_error()`.
+#define _ts_checked(status)                                       \
+  ({                                                              \
+    ts_t _checked = (status);                                     \
+    if ((_checked.code & 0xFFFF) != (_checked.code >> 16)) {      \
+      __fatal_error("ts_check() error", __FILE_NAME__, __LINE__); \
+    }                                                             \
+    _checked;                                                     \
+  })
```

Comment:
I wonder how much size and time overhead this will add throughout the entire code base and whether it's worth it. It could also be a barrier to optimization.

Reply:
I haven't looked at the assembly code yet, so I can't tell what the cost is.  That's just an idea and we can omit this code if we think it's not worth it. I think this could be also optional.